core/src/main/scala/kafka/server/KafkaServer.scala

```diff
@@ -322,6 +322,10 @@ class KafkaServer(
           autoTopicCreationChannel = Some(clientToControllerChannelManager)
         }
 
+        if (config.liDropCorruptedFilesEnable) {
+          zkClient.registerCorruptedBrokerId(config.brokerId)
+        }
```

Comment:
Other than the overall config, this appears to self-register all brokers as corrupted regardless of any actual data checks. Did I miss something? (Is this intended for a follow-on PR? Even if so, it's super-dangerous to allow it to go in early, unless this will be pushed to a separate branch...which doesn't seem to be the case according to the github UI.)

Reply:
This (and the issues noted in the above three paragraphs) were addressed in person today. The upshot is that SREs mark (via config) the specific brokers that are considered potentially corrupted (before any post-outage broker start, of course). Brokers that start up with that config then mark themselves on a ZK znode ("/corrupted" or something), which lets the controller know about them. The controller removes all of them from the ISRs of their respective partitions (for partitions where they're a follower), starts delayed elections for all partitions that have no uncorrupted replicas, blocks planned leader elections/etc. for the same, and then makes a set of ListOffsets requests to affected brokers in order to determine which one holds the latest data for each affected partition. Once the ListOffsets responses come back to the controller (or the timeout is reached), it then selects the best leader for each partition and completes the elections.

Note that if the brokers for some reason start without the "corrupted" config marking, they will do their normal, random leadership elections, and data will be lost. There's no recovery from this state, so it's important that runbooks are updated as needed to ensure other teams don't blindly restart Kafka servers when bringing up hosts (for example).